src/playback-watcher.js

```diff
@@ -161,15 +161,55 @@ export default class PlaybackWatcher {
    * @private
    */
   waiting_() {
+    if (this.techWaiting_()) {
+      return;
+    }
+
+    // All tech waiting checks failed. Use last resort correction
+    let currentTime = this.tech_.currentTime();
+    let buffered = this.tech_.buffered();
+    let currentRange = Ranges.findRange(buffered, currentTime);
+
+    // Sometimes the player can stall for unknown reasons within a contiguous buffered
+    // region with no indication that anything is amiss (seen in Firefox). Seeking to
+    // currentTime is usually enough to kickstart the player. This checks that the player
+    // is currently within a buffered region and there is at least half a second
+    // of forward buffer so that this isn't triggered when the player is just buffering
+    // due to slow connection.
+    if (currentRange.length && currentTime <= currentRange.end(0) - 0.5) {
+      this.cancelTimer_();
+      this.tech_.setCurrentTime(currentTime);
+
+      this.logger_(`Stopped at ${currentTime} while inside a buffered region ` +
+        `[${currentRange.start(0)} -> ${currentRange.end(0)}]. Attempting to resume ` +
+        `playback by seeking to the current time.`);
+
+      // unknown waiting corrections may be useful for monitoring QoS
+      this.tech_.trigger('unknownwaiting');
+      return;
+    }
+  }
+
+  /**
+   * Handler for situations when the tech fires a `waiting` event
```

Comment:
Would be good to add a short note about how we can't predict when these events will occur for the browser.